### PR TITLE
#308047: Changed the rule to composing mutex key to block simultaneous message

### PR DESCRIPTION
**Pull request by rafael-santiago-take**
Changed the rule to composing mutex key to block simultaneous message in subflow or router context

### CODE REVIEW

File: src/Take.Blip.Builder/FlowManager.cs [hunk 1]
```diff
@@ -157,7 +159,9 @@ public async Task ProcessInputAsync(Message message, Flow flow, IContext message
                 using (var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, cancellationToken))
                 {
                     // Synchronize to avoid concurrency issues on multiple running instances
-                    var handle = await _namedSemaphore.WaitAsync($"{flow.Id}:{userIdentity}", _configuration.InputProcessingTimeout, linkedCts.Token);
+                    // If the flow is contained in another flow, such as a subflow or a bot running on a router, the mutext key must consider the parent flow id.
```
Comment (caiopsouza) [inline src/Take.Blip.Builder/FlowManager.cs:162]: "mutex" is spelled "mutext" here.

File: src/Take.Blip.Builder/FlowManager.cs [hunk 2]
```diff
@@ -540,5 +544,25 @@ await output.Conditions.EvaluateConditionsAsync(lazyInput, context, cancellation
 
             return state;
         }
+
+        private string DefineMutexKey(Message message, Flow flow, Identity userIdentity)
+        {
+            string parent = flow.Id;
+            string identity = userIdentity.ToString();
```
Comment (caiopsouza) [inline src/Take.Blip.Builder/FlowManager.cs:551]: `var` is preferred.

File: src/Take.Blip.Builder/FlowManager.cs [hunk 3]
```diff
@@ -540,5 +544,25 @@ await output.Conditions.EvaluateConditionsAsync(lazyInput, context, cancellation
...
+            if (message.Metadata != null && message.Metadata.ContainsKey(TUNNEL_OWNER_METADATA))
+            {
+                //when the message is traveling in the context of a router, we must consider the key as the router identifier and the tunnel originator
+                message.Metadata.TryGetValue(TUNNEL_OWNER_METADATA, out parent);
+                message.Metadata.TryGetValue(TUNNEL_ORIGINATOR_METADATA, out identity);
+            }
+            else if (!flow.ParentId.IsNullOrEmpty())
+            {
+                //when the message is traveling in the context of subflow, we must consider the key as the parent flow
+                parent = flow.ParentId;
```
Comment (caiopsouza) [inline src/Take.Blip.Builder/FlowManager.cs:562]: Shouldn't the key be the parent plus the id of the flow. Wouldn't this have the same issue if a user is talking with a subflow in multiple bots?

File: src/Take.Blip.Builder/FlowManager.cs [hunk 4]
```diff
@@ -540,5 +544,25 @@ await output.Conditions.EvaluateConditionsAsync(lazyInput, context, cancellation
 
             return state;
         }
+
+        private string DefineMutexKey(Message message, Flow flow, Identity userIdentity)
+        {
+            string parent = flow.Id;
+            string identity = userIdentity.ToString();
+
+            if (message.Metadata != null && message.Metadata.ContainsKey(TUNNEL_OWNER_METADATA))
```
Comment (caiopsouza) [inline src/Take.Blip.Builder/FlowManager.cs:553]: Should it check for `TUNNEL_ORIGINATOR_METADATA` too? I understand it's expected to have both or none, but is it possible a malformed message would have only one?

File: src/Take.Blip.Builder/FlowManager.cs (hunk 3 above)
Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/FlowManager.cs:562]: within a conversation, the state is unique by bot...
If the user talks to the same subflow in differents bots, the key must be different too

File: src/Take.Blip.Builder/FlowManager.cs (hunk 1 above)
Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/FlowManager.cs]: done

File: src/Take.Blip.Builder/FlowManager.cs (hunk 2 above)
Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/FlowManager.cs]: done

**Review by jujucasseb — APPROVED** [review]

**Review by mhnakashima — APPROVED** [review]

**Review by caiopsouza — APPROVED** [review]

**Pull request merged**